fedorov/fedorov.py

```diff
@@ -202,6 +204,103 @@ def translate_to_vector_2D(a=1, b=1, theta=np.pi / 2):
     return lattice_vectors
 
 
+def populate_system(
+        lattice: np.ndarray,
+        basis_vectors: np.ndarray,
+        num_replicas: typing.Union[int, np.ndarray],
+        center: bool = True,
+        **kwargs: np.ndarray):
+    r"""Return positions and box dimensions for given number of replicas.
+
+    Computes the location for particle positions within the translated lattice
+    using the basis vectors. The number of replicas determines the number of
+    times the lattice is translated in each dimension. The function can also
+    expand any number of other properties associated with a basis vector
+    through key word arguments.
+
+    Parameters
+    ----------
+    lattice: (3, :math:`N_{dim}`) array of floats
+        The lattice vectors for the crystal.
+    basis_vectors: (:math:`N_b, N_{dim}` ) array of float
+        The basis vectors in fraction coordinates with respect to the given
+        lattice. These are the positions of particles within the lattice.
+    num_replicas: int or (:math:`N_{dim}`,) array of int
+        Either an array of the number of lattice relicas in each direction or an
+        integer which signifies replicating the unit cell ``num_replicas`` times
+        in each dimension.
+    center: bool, optional
+        Whether to roughly center positions. This is commonly done for
+        simulations such as those using HOOMD-blue, defaults to ``True``.
+    \*\*kwargs: np.ndarray
+        Optional key word arguments which each represent a property specific to
+        a particular basis vector (particle position) within the lattice. These
+        are use to create arrays that are the same length as the returned
+        positions with the same properties per basis vector as passed. For
+        instance, particle types can be computed this way.
+
+    Returns
+    -------
+    positions: (:math:`N_p`, 3) np.ndarray of float
+        An array of positions that are the result of translating the lattice
+        across space based on ``num_replicas``.
+    box: (6,) np.ndarray of float
+        The values for the containing periodic box,
+        :math:`L_x,\ L_y,\ L_z,\ xy, xz, yz`.
+    property_arrays: dict[str, np.ndarray], optional
+        A dictionary with keys from the key word arguments passed to the
+        function with NumPy arrays as values. Each NumPy array has been repeated
+        such that its relation to the basis vectors is preserved and the array
+        covers all positions. This is only returned with at least one keyword
+        argument is passed.
+    """
+    N_dims = basis_vectors.shape[-1]
+    if isinstance(num_replicas, int):
+        num_replicas = np.array((num_replicas,) * N_dims)
+
+    if len(num_replicas) != N_dims:
+        raise ValueError(
+            f"num_replicas must be an int or array of shape=({N_dims},)")
+
+    for key, arr in kwargs.items():
+        if not isinstance(arr, np.ndarray):
+            arr = np.array(arr)
+            kwargs[key] = arr
```

Comment:
Simplify. `asarray` is pretty inexpensive to call on array-like objects.
```suggestion
        arr = np.asarray(arr)
        kwargs[key] = arr
```